### ISSUE TITLE
Using replace to evaluate multiple gradients during training in Keras

### ISSUE THREAD

**Issue by slavakung** [Theano/Theano#5930]
I am a researcher in optimization and I am interested in testing algorithms for training DNNs using keras.

In practice, I would like to do something a bit different from the other optimizers, I would like to compute the gradient at a slightly different value of the tensor of parameters than the current one, and the update I will make to the parameters will depend on both the current gradient and this other gradient.

In practice this has proven more difficult than anticipated. See,

https://github.com/fchollet/keras/issues/6175
https://github.com/tensorflow/tensorflow/issues/9434

it appears this is not really possible with tensorflow. One responder suggested I should be able to do something like this using theano as the backend:

    def get_updates(self, params, constraints, loss):
        grads = self.get_gradients(loss, params)
        #tempparam = [tf.Variable(a) for a in params]
        tempparams = [a+1. for a in params]
        replace = {p:npm for p, npm in zip(params, tempparams)}
        gradsn = [th.replace(g, replace = replace) for g in grads]

but apparently theano has no function replace. Does Theano have any functionality to do this in the keras framework? Some have mentioned having to construct the graph again to have multiple evaluations of a gradient, but I am not sure what that even means in this context, if I want to perform a specific iteration within the get_updates call in keras. Is this at all possible with keras? Perhaps theano has some functionality like as the code above is intended to do I am not aware of?

If not, and it is not possible to compute more than one gradient during an optimization within keras, which somehow ties you to perform one forward and back pass with each step of the training optimization iteration, what would be the best way to go about this, in lieu of using keras? I am just looking to have a custom multi-layered perceptron, and perform training using an optimization procedure that calculates multiple gradients at every iteration. 

Thank you

**Comment by JesseLivezey** [comment]
Would ```theano.clone``` work?
http://deeplearning.net/software/theano/library/#theano.clone

For questions like this, you'll probably get better feedback on the google group
https://groups.google.com/forum/#!forum/theano-users
